authorize/evaluator/evaluator.go

```diff
@@ -4,48 +4,320 @@ package evaluator
+		return "", err
+	}
+
+	payload := map[string]interface{}{
+		"iss": e.authenticateHost,
+	}
+	if u, err := url.Parse(req.HTTP.URL); err == nil {
+		payload["aud"] = u.Hostname()
+	}
+	if s, ok := req.DataBrokerData.Get("type.googleapis.com/session.Session", req.Session.ID).(*session.Session); ok {
+		if tm, err := ptypes.Timestamp(s.GetIdToken().GetExpiresAt()); err == nil {
+			payload["exp"] = tm.Unix()
+		}
+		if tm, err := ptypes.Timestamp(s.GetIdToken().GetIssuedAt()); err == nil {
+			payload["iat"] = tm.Unix()
+		}
+		if u, ok := req.DataBrokerData.Get("type.googleapis.com/user.User", s.GetUserId()).(*user.User); ok {
+			payload["sub"] = u.GetId()
+			payload["email"] = u.GetEmail()
+		}
+		if du, ok := req.DataBrokerData.Get("type.googleapis.com/directory.User", s.GetUserId()).(*directory.User); ok {
+			payload["groups"] = du.GetGroups()
+		}
+	}
+
+	bs, err := json.Marshal(payload)
+	if err != nil {
+		return "", err
+	}
+
+	jws, err := signer.Sign(bs)
+	if err != nil {
+		return "", err
+	}
+
+	return jws.CompactSerialize()
+}
+
+type input struct {
+	DataBrokerData           DataBrokerData `json:"databroker_data"`
+	HTTP                     RequestHTTP    `json:"http"`
+	Session                  RequestSession `json:"session"`
+	IsValidClientCertificate bool           `json:"is_valid_client_certificate"`
+}
+
+func (e *Evaluator) newInput(req *Request, isValidClientCertificate bool) *input {
+	i := new(input)
+	i.DataBrokerData = req.DataBrokerData
+	i.HTTP = req.HTTP
+	i.Session = req.Session
+	i.IsValidClientCertificate = isValidClientCertificate
+	return i
+}
+
+type (
+	// Request is the request data used for the evaluator.
+	Request struct {
+		DataBrokerData DataBrokerData `json:"databroker_data"`
+		HTTP           RequestHTTP    `json:"http"`
+		Session        RequestSession `json:"session"`
+	}
+
+	// RequestHTTP is the HTTP field in the request.
+	RequestHTTP struct {
+		Method            string            `json:"method"`
+		URL               string            `json:"url"`
+		Headers           map[string]string `json:"headers"`
+		ClientCertificate string            `json:"client_certificate"`
+	}
+
+	// RequestSession is the session field in the request.
+	RequestSession struct {
+		ID                string   `json:"id"`
+		ImpersonateEmail  string   `json:"impersonate_email"`
+		ImpersonateGroups []string `json:"impersonate_groups"`
+	}
+)
+
+// Result is the result of evaluation.
+type Result struct {
+	Status    int
+	Message   string
+	SignedJWT string
+}
+
+func getAllowVar(vars rego.Vars) bool {
+	result, ok := vars["result"].(map[string]interface{})
+	if !ok {
+		return false
+	}
+
+	allow, ok := result["allow"].(bool)
+	if !ok {
+		return false
+	}
+	return allow
+}
+
+func getDenyVar(vars rego.Vars) []Result {
+	result, ok := vars["result"].(map[string]interface{})
+	if !ok {
+		return nil
+	}
+
+	denials, ok := result["deny"].([]interface{})
+	if !ok {
+		return nil
+	}
+
+	var results []Result
```

Comment:
Nit, we can do:

```
results := make([]Result, 0, len(denials))
```

since when we now `len(results)` will always less than or equal to `len(denials)`. This helps appending to `results` better performance.